ownership-chain/pallets/asset-metadata-extender/src/tests.rs

```diff
@@ -0,0 +1,134 @@
+	new_test_ext().execute_with(|| {
+		// Go past genesis block so events get deposited
+		System::set_block_number(1);
+		let claimer = H160::zero();
+		let universal_location: BoundedVec<u8, MaxUniversalLocationLength> = bounded_vec![0; 1];
+		let token_uri: BoundedVec<u8, MaxTokenUriLength> = bounded_vec![0; 1];
+
+		create_metadata_extension(claimer, universal_location.clone(), token_uri.clone());
+
+		System::assert_last_event(
+			Event::MetadataExtensionCreated { universal_location, claimer, token_uri }.into(),
+		);
+	});
+}
+
+#[test]
+fn given_an_ul_and_token_uri_i_cannot_create_twice_asset_extension_with_same_claimer() {
+	new_test_ext().execute_with(|| {
+		let claimer = H160::zero();
+		let universal_location: BoundedVec<u8, MaxUniversalLocationLength> = bounded_vec![0; 1];
+		let token_uri: BoundedVec<u8, MaxTokenUriLength> = bounded_vec![0; 1];
+
+		create_metadata_extension(claimer, universal_location.clone(), token_uri.clone());
+		assert_noop!(
+			AssetMetadataExtender::create_metadata_extension(
+				claimer,
+				universal_location,
+				token_uri
+			),
+			Error::<Test>::MetadataExtensionAlreadyExists
+		);
+	});
+}
+
+#[test]
+fn given_an_universal_location_after_creating_extension_counter_increases() {
+	new_test_ext().execute_with(|| {
+		let claimer = H160::zero();
+		let universal_location: BoundedVec<u8, MaxUniversalLocationLength> = bounded_vec![0; 1];
+		let token_uri: BoundedVec<u8, MaxTokenUriLength> = bounded_vec![0; 1];
+
+		// create first extension for the given UL
+		assert_eq!(
+			AssetMetadataExtender::metadata_extensions_counter(universal_location.clone()),
+			0
+		);
+
+		create_metadata_extension(claimer, universal_location.clone(), token_uri.clone());
+		assert_eq!(
+			AssetMetadataExtender::metadata_extensions_counter(universal_location.clone()),
+			1
+		);
+
+		// check that no other UL has been affected
+		let another_universal_location: BoundedVec<u8, MaxUniversalLocationLength> =
+			bounded_vec![1; 1];
+		assert_eq!(
+			AssetMetadataExtender::metadata_extensions_counter(another_universal_location),
+			0
+		);
+
+		// create another extension for the same UL with another claimer
+		let another_claimer = H160::from_low_u64_be(1);
+		create_metadata_extension(another_claimer, universal_location.clone(), token_uri);
+		assert_eq!(AssetMetadataExtender::metadata_extensions_counter(universal_location), 2);
+	});
+}
+
+#[test]
+fn given_an_ul_i_can_get_all_its_extensions() {
+	new_test_ext().execute_with(|| {
+		let universal_location: BoundedVec<u8, MaxUniversalLocationLength> = bounded_vec![0; 1];
+		let token_uri: BoundedVec<u8, MaxTokenUriLength> = bounded_vec![0; 1];
+
+		let n = 1000;
+		for i in 0..n {
+			let claimer = H160::from_low_u64_be(i);
+			create_metadata_extension(claimer, universal_location.clone(), token_uri.clone());
+		}
+
+		for i in 0..n {
+			assert_eq!(
+				AssetMetadataExtender::metadata_extensions(universal_location.clone(), i)
```

Comment:
I was confused by the implementation of this test till I recognized the pattern. Would be more efficient and readable to get the value from metadata_extensions, then check its elements ?

Reply:
The method `metadata_extensions` returns one element at a time according to the index. I'm not sure about you are proposing, do you want to add another for so that I get all the elements in the existing one and read all the elements in the new added one?

Reply:
it not very important cause it's a test but would it be better to call `AssetMetadataExtender::metadata_extensions` once in this loop ?

Reply:
it's not very important cause it's a test but would it be nicer if `metadata_extensions` is called once in the loop ?